rfcs/1440_generic_planner_node.md

```diff
@@ -0,0 +1,237 @@
+
+## Background
+
+An Oak application creates an initial node and calls the entry point on that
+node. A typical pattern is to use this initial node as a "Planner Node",
+responsible for creating all other nodes and channels. The creation of the nodes
+an channels is done through code. For simple cases it is easy to figure out what
+the graph can look like by inspecting the code, but it becomes increasingly
+difficult as the complexity of the logic of the planner node increases to cope
+with more complex node graphs. Another drawback of the code-based dynamic node
+creation is that we have to run the code to use the introspection mechanism to
+visualise the node graph.
+
+Generating the node graph automatically based on a configuration/manifest file
+should simplify this. It should be easier to figure out what the node graph will
+look like through inspecting a manifest file (assuming it has an easily
+understandable format). Additional benefits of configuration-based node creation
+include:
+
+- It would be easier to build tools to visualise the node graph without running
+  the application.
+- It would be easier to build tools to perform static analysis of the node graph
+  to detect potential problems.
+- It would usualy be easier to reason about the security of a static node graph.
+
+## Overview
+
+Moving all node and channel creation to a configuration-based mechanism is a
+significant and potentially disruptive change. This RFC proposes a more gradual
+approach by implementing a re-usable WASM-based planner node that uses a
+manifest file to do the node and channel creation. The functionality can be
+expanded over time to support more complex scenarios.
+
+The planner node will be responsible for creating nodes, creating channels,
+sending channel handles to the created nodes to ensure that the full node graph
+is created on start-up. It will will also act as a message router, which would
+route HTTP- and GRPC invocations to other nodes based on the associated labels.
+
+## Detailed Design
+
+### Planner Node
+
+The planner node will the the initial node created for an application. The
+manifest file will be passed to the planner node as part of the application's
+`ConfigMap`.
+
+The planner node will create all the nodes and channels as specified with the
+appropriate labels. It will send initialisation messages to each of the nodes,
+providing additional configuration and handles to channels. It is the
+responsibility of each of the nodes to store these channels for future
+communications with other nodes. It will send a filtered version of the config
+map as part of the initialisation message, based on the configuration keys
+specified in the `config_files` field.
+
+### Invocation Routing
+
+The planner node will also be responsible for routing invocations from the gRPC-
+and HTTP server pseudo nodes to the right receiver node based on the associated
+labels. Since it would be possible for multiple nodes to have the same Label,
```

Comment:
nit: inconsistent capitalization of "Label"

Reply:
Done.